server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -550,22 +550,39 @@ public String toString() {
             this.absoluteMillis = System.currentTimeMillis();
             this.counter = new TimeCounter();
             setDaemon(true);
+            if (interval <= 0) {
+                this.running = false;
+            }
         }
 
         /**
          * Return the current time used for relative calculations. This is
          * {@link System#nanoTime()} truncated to milliseconds.
+         * <p>
+         * If {@link ThreadPool#ESTIMATED_TIME_INTERVAL_SETTING} is set to 0
+         * then the cache is disabled and the method calls {@link System#nanoTime()}
+         * whenever called. Typically used for testing.
          */
         long relativeTimeInMillis() {
-            return relativeMillis;
+            if (running) {
```

Comment:
I'd slightly prefer

```suggestion
            if (0 < interval) {
```

since this is what the Javadoc says.

Reply:
I chose to use the boolean `running` (properly set in the constructor) to avoid the comparison on every call, and it's used in the same way to avoid having a "running" thread that updates those values.

Reply:
I don't clearly understand why avoiding the comparison is a good thing. Is it a performance question? Are you sure that a comparison between a constant and a `final` field is worse than a second `volatile` read?

Reply:
My thinking was yes the performance, but didn't think about the volatile.
But also to have the logic that when `interval == 0` -> `running = false` so that we avoid having a running thread and we have a common check in all 3 places (the methods and the `run()`).

If you think that the interval check is better, I'll happily change. And if we do that, should the code in the run() change like: `while(running && 0 < interval)` ?

Reply:
Yep, sounds like a good idea, thanks.